### ISSUE TITLE
System tests fail to run on Cygwin environment 

### ISSUE THREAD

**Issue by Mesbah-Alam** [AdoptOpenJDK/openjdk-systemtest#165]
Systemtest fail to run on Cygwin environment that is offered by the Windows cloud machines. They fail with the following error:

```"SharedClassesWorkloadTest_Softmx_Increase_0""_FAILED"; echo ""; fi; } 2>&1 | tee -a "C:/Users/jenkins/workspace/Grinder/openjdk-tests/\\TestConfig\\test_output_1535039567450\\TestTargetResult";
11:52:52 Can't locate stf/stfUtility.pm in @INC (you may need to install the stf::stfUtility module) (@INC contains: /cygdrive/c/Users/jenkins/workspace/Grinder /usr/local/lib/perl5/site_perl/5.26/x86_64-cygwin-threads /usr/local/share/perl5/site_perl/5.26 /usr/lib/perl5/vendor_perl/5.26/x86_64-cygwin-threads /usr/share/perl5/vendor_perl/5.26 /usr/lib/perl5/5.26/x86_64-cygwin-threads /usr/share/perl5/5.26) at C:/Users/jenkins/workspace/Grinder/openjdk-tests///..//jvmtest\systemtest\stf\stf.core\scripts\stf.pl line 30.
11:52:52 BEGIN failed--compilation aborted at C:/Users/jenkins/workspace/Grinder/openjdk-tests///..//jvmtest\systemtest\stf\stf.core\scripts\stf.pl line 30.
```

https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/367/console
Perl version used in Cygwin : 

```
$ perl --version

This is perl 5, version 26, subversion 2 (v5.26.2) built for x86_64-cygwin-threads-multi
(with 7 registered patches, see perl -V for more detail)

Copyright 1987-2018, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.

```

make version used : 
```
$ make --version
GNU Make 4.2.1
Built for x86_64-unknown-cygwin
Copyright (C) 1988-2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

`stf.core` lists Strawberry Perl as a [preference ](https://github.com/AdoptOpenJDK/openjdk-systemtest/blob/master/openjdk.build/docs/build.md) . We need to investigate if that's causing the tests to fail.

**Comment by Mesbah-Alam** [comment]
Fyi: @lumpfish @smlambert @llxia

**Comment by llxia** [comment]
System test on wins are excluded in OpenJ9 Jenkins nightly runs as almost all the system tests failed with above error.

https://ci.eclipse.org/openj9/view/Test/job/Test-extended.system-JDK10-win_x86-64_cmprssptrs/

Once system test is fixed on wins, we need to reenable them in OpenJ9 Jenkins nightly build.
fyi @AdamBrousseau

**Comment by Mesbah-Alam** [comment]
The error seen when system test is run at Adopt on Windows, is: 

```
ClassLoadingTest_0 Start Time: Tue Nov  6 16:33:51 2018 Epoch Time (ms): 1541522031994
test with NoOptions
Can't locate stf/stfUtility.pm in @INC (you may need to install the stf::stfUtility module) (@INC contains: /cygdrive/c/Users/jenkins/workspace/Grinder /usr/local/lib/perl5/site_perl/5.26/x86_64-cygwin-threads /usr/local/share/perl5/site_perl/5.26 /usr/lib/perl5/vendor_perl/5.26/x86_64-cygwin-threads /usr/share/perl5/vendor_perl/5.26 /usr/lib/perl5/5.26/x86_64-cygwin-threads /usr/share/perl5/5.26) at C:/Users/jenkins/workspace/Grinder/openjdk-tests///..//jvmtest\systemtest\stf\stf.core\scripts\stf.pl line 30.
BEGIN failed--compilation aborted at C:/Users/jenkins/workspace/Grinder/openjdk-tests///..//jvmtest\systemtest\stf\stf.core\scripts\stf.pl line 30.
```
https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/367/tapResults/

@lumpfish, Just to refresh memories here, we are seeing the above error on all stf based system tests when run on the Adopt cloud Windows machines, which run on Cygwin make. Could you please share your thoughts on how we should approach getting `stfUtility.pm` working on Cygwin? Is there any dependency of `stfUtility.pm` on Strawberry Perl which is causing it to fail on Cygwin?

**Comment by smlambert** [comment]
That error is actually just complaining that the perl module is not found, have we tried hacking the scripts to copy that stfUtility.pm into one of the locations the error msg suggests it could be found, or add a the directory where the .pm file is to the list of dirs in @INC ?

**Comment by lumpfish** [comment]
This is the code which should enable stfUtility to be located (in stf.pl):
```
use FindBin qw($Bin);
use lib "$Bin";
use stf::stfUtility;
```
It says 'Find the directory the current script (stf.pl) is in and add it to the library list (i.e. add it to @INC)'.  So that appears to be not working correctly on cygwin perl.

I would suggest reproducing the error by running a test on a local machine with cygwin installed from a cygwin command prompt and with no Strawberry perl on the path.  So for this first error, print out what FindBin is returning.  That might help identify where the problem lies (a 'weird' path spec?).

It may be that there are other issues running with cygwin perl.  There is also the danger that any 'fixes' to make the tests run with cygwin perl might prevent them running with Strawberry perl.

The fastest fix may well be to install Strawberry perl and put it ahead of cygwin perl in the cygwin PATH before the tests are started.

**Comment by Mesbah-Alam** [comment]
Waiting on Infra team's update about using fire machines-- which would by default have Strawberry Perl installed.

**Comment by smlambert** [comment]
Agreed, if the majority of the machines that these tests will be run on have Strawberry perl installed on them by default, then it would make sense to install on the remainder of machines and live with the additional 'pre-req'.

**Comment by lumpfish** [comment]
I've done some investigation locally.
There are two areas of issues:

cygwin shell vs cmd.exe
- the makefiles currently force the use of cmd.exe using the gmake SHELL macro.  However, cygwin commands such as rmdir or dir are found ahead of the Windows commands which do not behave the same
- powershell commands hang running under cygwin shell
- make target command lines which set environment variables must use export if using the cygwin shell and set if running under cmd.exe.  Currently there is an EXPORT make macro with the value of 'set' on Windows (and therefore on cygwin).

These solutions work in my testing locally:
a. For all except the powershell issue, prefixing all Windows commands with "cmd / c" when running under cygwin, rather than setting the SHELL environment
b. For the powershell hang problem, the suggestion here https://stackoverflow.com/questions/6333276/how-to-end-s-powershell-script-when-running-from-cygwin also works in my testing (```echo \n | powershell xxxxx```)
c. For the environment value issue, setting the EXPORT macro to 'export' rather than 'set' when running under cygwin.

2. Issues due to cygwin perl vs. Strawberry perl
- These are numerous.  The main issue is path names. The perls recognise different syntax for paths.  Also any Windows commands issued from inside perl scripts require the paths to be in Windows format.   It may be that there is a way to get all the tests and all the infrastructure to format for either, but it would not be easy, or quick.
So the easiest solution is to install Strawberry perl on the Windows test machine.
I have added some code to find a Windows port of perl in the PATH (by looking for perl.exe in a perl/bin directory which is not where it is in cygwin) and if found move it (actually add it) to the front of PATH, so it will be found before cygwin perl.  The tests will stop if no such perl can be found with an error message.

- So I have tested with Strawberry perl in the following setups
-- Windows command prompt, win32 make port (i386-pc-mingw32)
-- cygwin terminal, win32 make port (i386-pc-mingw32), Strawberry perl ahead of cygwin perl on the PATH
-- cygwin terminal, cygwin make, , Strawberry perl ahead of cygwin perl on the PATH

There are three repositories which need changing / testing: stf, openjdk-systemtest and openj9-systemtest (at eclipse/openj9).

@Mesbah-Alam  - please could you try some runs with the changes and see if it fixes the issues you've been seeing?

These are the PRs:
https://github.com/AdoptOpenJDK/stf/pull/48
https://github.com/AdoptOpenJDK/openjdk-systemtest/pull/195
https://github.com/eclipse/openj9-systemtest/pull/56

**Comment by lumpfish** [comment]
I also added a new makefile target test.java8 in two of the PRs to avoid running the new tests which have been added to the make test target but cannot run against jdk8 level code (e.g. modularity tests).

**Comment by Mesbah-Alam** [comment]
@lumpfish - I tried your changes in a grinder build in Adopt and it still did not work unfortunately : https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/451/console

We still see all the tests failing with : 

```
Can't locate stf/stfUtility.pm in @INC (you may need to install the stf::stfUtility module) (@INC contains: /cygdrive/c/Users/jenkins/workspace/Grinder /usr/local/lib/perl5/site_perl/5.26/x86_64-cygwin-threads /usr/local/share/perl5/site_perl/5.26 /usr/lib/perl5/vendor_perl/5.26/x86_64-cygwin-threads /usr/share/perl5/vendor_perl/5.26 /usr/lib/perl5/5.26/x86_64-cygwin-threads /usr/share/perl5/5.26) at C:/Users/jenkins/workspace/Grinder/openjdk-tests///..//jvmtest\systemtest\stf\stf.core\scripts\stf.pl line 30.
BEGIN failed--compilation aborted at C:/Users/jenkins/workspace/Grinder/openjdk-tests///..//jvmtest\systemtest\stf\stf.core\scripts\stf.pl line 30.
```

I logged into the machine manually to double check if your changes were cloned from your branch properly, and they were.

**Comment by lumpfish** [comment]
"I logged into the machine manually to double check if your changes were cloned from your branch properly, and they were."
I can't see where in the job that clone happens?

**Comment by lumpfish** [comment]
Ah - the reason you're not seeing the changes may be because you're not executing them!  Are you running the openjdk-systemtest makefile in your testkitgen setup?  That's where most of the changes are.  If you are bypassing that makefile you will need to put similar changes into whatever makefile wrapper you have replaced them with.  To summarise the changes are:
1. Detect whether you are running under cygwin
2. If so
a. Prefix commands with cmd /c
b. Detect whether a Windows perl port is present on the path
-- If so, add it to front of PATH 
-- If not, output an error message and stop
c. If you are setting environment variables in the make command lines you need to use 'export' if you're in a cygwin environment and 'set' if you're not.
Or you can run make test.xxxx from makefiles in the test repos which will do all that for you.
Note that you will still need the PRs above for people running tests directly from the repos and to pick up the other changes in stf.pl and Commands.pm which also have fixes to allow them to run under cygwin.

**Comment by Mesbah-Alam** [comment]
I am cloning your branches (there are 3 of them) via the systemtest/build.xml script at Adopt : https://github.com/Mesbah-Alam/openjdk-tests/blob/trySimonsFixForWin/systemtest/build.xml 

Yes, the job doesn't echo explicitly which branch  it is cloning. That's why you couldn't see that info in the console output.

**Comment by Mesbah-Alam** [comment]
```
Ah - the reason you're not seeing the changes may be because you're not executing them! Are you running the openjdk-systemtest makefile in your testkitgen setup? That's where most of the changes are.
```

- This could be the issue. @renfeiw - can you confirm if we bypass the [openjdk-systemtest makefile ](https://github.com/AdoptOpenJDK/openjdk-systemtest/blob/master/openjdk.build/makefile) when we run system tests via testkitgen?

**Comment by Mesbah-Alam** [comment]
[systemtest.mk](https://github.com/Mesbah-Alam/openjdk-tests/blob/addSystemTestMakefile/systemtest/systemtest.mk) file containing the logic mentioned in [#165 (comment)](https://github.com/AdoptOpenJDK/openjdk-systemtest/issues/165#issuecomment-441598524) has been added. 

However, we now have a new issue : 
```
13:11:25 systemtest.mk:92: *** Unable to find Windows perl e.g. Strawberry perl in a /perl/bin subdirectory on PATH.  Install perl or add to PATH and retry.  Stop.
```
https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/457/console 

The logic in systemtest.mk that looks for perl and gives the above error is : 

```
PERL:=$(shell which -a perl.exe | grep /perl/bin | sort | uniq)
    ifeq (,$(PERL))
      $(error Unable to find Windows perl e.g. Strawberry perl in a /perl/bin subdirectory on PATH.  Install perl or add to PATH and retry)
    else
      $(warning Found perl in $(PERL))
endif
```
Hi @sxa555 , do we have Perl installed on https://ci.adoptopenjdk.net/computer/test-azure-win2012r2-x64-1/ ?

**Comment by Mesbah-Alam** [comment]
@sxa555 ensured that`/cygdrive/c/Strawberry/perl/bin` is in the PATH (and and ahead of the cygwin Perl) on the Adopt test Windows machines. This ensures that, with Simon's make changes added to `systemtest.mk`, we have the System tests running on Windows at Adopt: https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/472/consoleFull

**Comment by sxa555** [comment]
For future reference, Strawberry Perl was installed on all the machines but the `PATH` had an entry for `C:\Strawberry\bin` instead of `C:\Strawberry\perl\bin` (Or somethign very similar!) I fixed it on the two SoftLayer Wundows Machines and the Azure one. The packet machine appeared to have the correct `PATH` entry already in place.

**Comment by AdamBrousseau** [comment]
And fixed via AdoptOpenJDK/infrastructure#603

**Comment by Mesbah-Alam** [comment]
Closing the issue since the problem has been fixed as mentioned above.
